File: backend/sla.py
from datetime import datetime, timedelta

# SLA window per priority, computed from created_at rather than persisted —
# the deadline is always derivable and shouldn't drift from whatever the
# priority happens to be at read time.
SLA_WINDOWS: dict[str, timedelta] = {
    "P0": timedelta(hours=2),
    "P1": timedelta(hours=12),
    "P2": timedelta(days=7),
    "P3": timedelta(days=14),
}


def sla_deadline(priority: str, created_at: datetime) -> datetime:
    return created_at + SLA_WINDOWS[priority]


def is_breached(priority: str, created_at: datetime, resolved_at: datetime | None, now: datetime) -> bool:
    """True if the SLA deadline passed before resolution (or before `now`,
    for a request that isn't resolved yet)."""
    deadline = sla_deadline(priority, created_at)
    reference_time = resolved_at if resolved_at is not None else now
    return reference_time > deadline

File: frontend/src/lib/sla.js
// Must stay in sync with backend/sla.py — SLA window per priority, computed
// from created_at rather than persisted.
export const SLA_WINDOWS_MS = {
  P0: 2 * 60 * 60 * 1000,
  P1: 12 * 60 * 60 * 1000,
  P2: 7 * 24 * 60 * 60 * 1000,
  P3: 14 * 24 * 60 * 60 * 1000,
}

// "Approaching" once 25% or less of the window remains.
const WARNING_THRESHOLD = 0.25

// The backend sends naive timestamps (no trailing Z/offset) that are always
// UTC (see backend/timeutils.py). A bare `new Date("2026-...T...")` parses
// that as *local* time instead, silently shifting the deadline by the
// viewer's UTC offset — append Z so it's parsed as UTC like it actually is.
function parseUtc(value) {
  if (value instanceof Date) return value
  const hasTimezone = /Z|[+-]\d\d:\d\d$/.test(value)
  return new Date(hasTimezone ? value : `${value}Z`)
}

export function getSlaStatus(priority, createdAt, now = new Date()) {
  const windowMs = SLA_WINDOWS_MS[priority]
  const created = parseUtc(createdAt)
  const deadline = new Date(created.getTime() + windowMs)
  const remainingMs = deadline.getTime() - now.getTime()

  let status
  if (remainingMs <= 0) {
    status = 'breached'
  } else if (remainingMs <= windowMs * WARNING_THRESHOLD) {
    status = 'warning'
  } else {
    status = 'ok'
  }

  return { deadline, remainingMs, status }
}

function formatDuration(ms) {
  const abs = Math.abs(ms)
  const minutes = Math.round(abs / 60000)
  if (minutes < 60) return `${minutes}m`
  const hours = Math.round(minutes / 60)
  if (hours < 24) return `${hours}h`
  const days = Math.round(hours / 24)
  return `${days}d`
}

export function formatSlaRemaining({ remainingMs, status }) {
  if (status === 'breached') return `Breached by ${formatDuration(remainingMs)}`
  return `${formatDuration(remainingMs)} left`
}

// For an already-resolved request: a fixed past outcome, not a live
// countdown — "3h left" would be misleading once the ticket is closed.
export function formatSlaOutcome(remainingMs) {
  return remainingMs >= 0 ? 'Met SLA' : `Missed SLA by ${formatDuration(remainingMs)}`
}
